### PR TITLE
Adds network_type to bookings

### LINKED ISSUE #25: Bookings - Alterations
1. Bookings should be able to have a "network_type"
* One of: Network or Sweatbox

2. Bookings should be able to have multiple positions associated with them

3. Booking logic must be altered to adapt to the above
* When creating a network booking (any user), it should check for other network bookings (and vice versa with sweatbox)

### PR DESCRIPTION
**Pull request by keanucz**
Fixes #25, adds a network_type column with tests to make sure different bookings of different network types can overlap and whatnot, and this better categorises the CTS modules into their own directories.

### CODE REVIEW

**Comment by keanucz** [comment]
@atoff Have a look :)

File: app/Modules/Bookings/Booking.php [hunk 1]
```diff
@@ -6,13 +6,13 @@
...
+use VATSIMUK\Support\Auth\Models\Concerns\HasRemoteRelationships;
 
 class Booking extends Model
 {
-    use HasCustomInstanceCreation;
+    use HasRemoteRelationships;
 
-    protected $fillable = ['user_id', 'position_id', 'from', 'to'];
+    protected $fillable = ['user_id', 'position_id', 'from', 'to', 'network_type'];
     protected $dates = ['from', 'to'];
```
Comment (atoff) [inline app/Modules/Bookings/Booking.php:20]: Instead of hard coding 0 as network and 1 as sweatbox, it might be better to use constants.  Define them here, like so:
```php
const NETWORK_TYPE_LIVE = 0;
const NETWORK_TYPE_SWEATBOX = 1;
```

and then refer to these later on in the logic

File: app/Modules/Bookings/Services/BookingsService.php [hunk 2]
```diff
@@ -19,6 +22,22 @@ public function __construct(User $user)
...
+    /**
+     * Check whether we've got a valid network type on our hands.
+     *
+     * @param  string  $type
+     * @return bool
+     */
+    public function validateNetworkType(int $type): bool
+    {
+        // 0 = live network, 1 = sweatbox
+        if ($type == 0 || $type == 1) {
```
Comment (atoff) [inline app/Modules/Bookings/Services/BookingsService.php]: As above, use the defined constants here. Would recommend we instead use something like:
```php
if(in_array($type, [Booking::NETWORK_TYPE_LIVE, Booking::NETWORK_TYPE_SWEATBOX])) {
```
here maybe? Could also instead define the array of possible network types on the model.

You could also inline this whole statement to:
```php
return in_array($type, [Booking::NETWORK_TYPE_LIVE, Booking::NETWORK_TYPE_SWEATBOX]);
```

File: composer.json [hunk 3]
```diff
@@ -15,8 +15,8 @@
         "laravel/telescope": "^2.0",
         "laravel/tinker": "^1.0",
         "mll-lab/laravel-graphql-playground": "^1.1",
-        "vatsimuk/auth-package": "dev-master",
-        "nuwave/lighthouse": "^4.2"
+        "nuwave/lighthouse": "^4.2",
+        "vatsimuk/auth-package": "dev-fix/relationship-mocks"
```
Comment (atoff) [inline composer.json]: This branch doesn't exist anymore. Can you change to re-target master, and update .lock to match this?

File: database/factories/BookingFactory.php [hunk 4]
```diff
@@ -8,11 +8,13 @@
...
     $from = $faker->dateTimeThisYear;
     $to = $faker->dateTimeInInterval($from, '+1 day');
+    $network_type = 0;
 
     return [
         'user_id' => $faker->numberBetween(1000000, 9999999),
         'position_id' => factory(\App\Modules\Position\Position::class),
         'from' => $from,
         'to' => $to,
+        'network_type' => $network_type,
```
Comment (atoff) [inline database/factories/BookingFactory.php:18]: You can inline this aswell (below assumes the using the constants suggested above)
```suggestion
        'network_type' => Booking::NETWORK_TYPE_LIVE,
```

File: docker-compose.yml [hunk 5]
```diff
@@ -11,7 +11,7 @@ services:
     networks:
       - app-network
     ports:
-      - 33060:3306 # expose db to local machine on port 33060
```
Comment (atoff) [inline docker-compose.yml]: @AxonC  was there a reason this was like this?

File: graphql/bookings.graphql [hunk 6]
```diff
@@ -15,18 +16,20 @@ input UpdateBookingInput {
...
+    createBooking(input: CreateBookingInput! @spread): Booking! @field(resolver: "App\\Modules\\Bookings\\Mutations\\CreateBookingHandler")
+    updateBooking(input: UpdateBookingInput! @spread): Booking! @field(resolver: "App\\Modules\\Bookings\\Mutations\\UpdateBookingHandler")
     deleteBooking(id: ID!): Booking! @delete
 }
 
 type Booking implements Node {
     id: ID!
     user_id: ID!
     position: Position! @belongsTo
+    network_type: Int!
```
Comment (atoff) [inline graphql/bookings.graphql]: Might be worth making this an enum, check out https://lighthouse-php.com/4.12/the-basics/types.html#enum

File: tests/Feature/Booking/BookingCreateTest.php [hunk 7]
```diff
@@ -64,6 +64,7 @@ public function testValidBookingCanBeCreated()
             createBooking(
                 input: {
                     position_id: {$this->position->id},
+                    network_type: 0,
```
Comment (atoff) [inline tests/Feature/Booking/BookingCreateTest.php]: Convert these to the constants defined above for readability

File: tests/Feature/Booking/BookingUpdateTest.php [hunk 8]
```diff
@@ -35,6 +35,7 @@ public function testItCanUpdateABooking()
             'position_id' => 1,
             'from' => new Carbon('2019-08-20 15:00:00'),
             'to' => new Carbon('2019-08-20 16:30:00'),
+            'network_type' => 0,
```
Comment (atoff) [inline tests/Feature/Booking/BookingUpdateTest.php]: As above

File: tests/Unit/Booking/BookingServiceUpdateTest.php [hunk 9]
```diff
@@ -71,28 +72,89 @@ public function itChecksForOverlappingBookingsOnUpdate()
         // explicitly setting booking times to stop coincidental test pass
         $this->existingBooking->from = new Carbon('10th January 2019 13:30:00');
         $this->existingBooking->to = new Carbon('10th January 2019 15:30:00');
+        $this->existingBooking->network_type = 0;
```
Comment (atoff) [inline tests/Unit/Booking/BookingServiceUpdateTest.php]: As above

**Review by atoff — APPROVED** [review]
Looks good! A few small things to iron out and possible ways to improve the code. Let me know what you think or if you have any questions :)

**Review by atoff — CHANGES_REQUESTED** [review]
Accidentally approved the last one, whoops

File: app/Modules/Bookings/Booking.php (hunk 1 above)
Comment (keanucz) [reply app/Modules/Bookings/Booking.php:20]: Honestly, this looks even nicer. I'm not very experienced with Laravel so this helps a lot

File: app/Modules/Bookings/Services/BookingsService.php (hunk 2 above)
Comment (keanucz) [reply app/Modules/Bookings/Services/BookingsService.php]: Yeah, definitely.

File: composer.json (hunk 3 above)
Comment (keanucz) [reply composer.json]: Hrm, not sure what happened there. I'll be sure to fix that

File: docker-compose.yml (hunk 5 above)
Comment (keanucz) [reply docker-compose.yml]: I think I had issues with docker-compose, this seemed to fix it for me.

File: graphql/bookings.graphql (hunk 6 above)
Comment (keanucz) [reply graphql/bookings.graphql]: Interesting, will look into it.

**Comment by keanucz** [comment]
Also, any idea why codecov isn't working?

**Comment by atoff** [comment]
> Also, any idea why codecov isn't working?

Doesn't seem to work outside of the main vatsim-uk repo for some reason. I think @AxonC was going to see if we could fix this?

File: app/Modules/Endorsement/Special/Services/SpecialEndorsementAssignment.php [hunk 10]
```diff
@@ -22,11 +22,11 @@ public function __construct(EndorsementRequest $endorsement, User $actioner)
     public function handle()
     {
         // check if the endorsement has already been assigned.
-        if ($this->endorsementRequest->endorsement->users->contains($this->endorsementRequest->user)) {
+        if ($this->endorsementRequest->endorsement->users()->contains($this->endorsementRequest->user_id)) {
```
Comment (AxonC) [inline app/Modules/Endorsement/Special/Services/SpecialEndorsementAssignment.php:25]: Why was this change needed?

File: docker-compose.yml (hunk 5 above)
Comment (AxonC) [reply docker-compose.yml]: This means you have a conflict locally on the port 33060 - please revert the change.

File: tests/Unit/Availability/AvailabilityUnitTest.php [hunk 11]
```diff
@@ -24,11 +24,6 @@ protected function setUp(): void
     /** @test */
     public function itReturnsTheUserRelation()
     {
-        $this->mock(User::class, function ($mock) {
```
Comment (AxonC) [inline tests/Unit/Availability/AvailabilityUnitTest.php:27]: Why is this removed? Without it, I believe it will call an actual API.

File: tests/Unit/Endorsement/SpecialEndorsementRequestServiceTest.php [hunk 12]
```diff
@@ -86,15 +86,21 @@ public function itCanActionRequestsForSpecialEndorsements()
         $this->mock(User::class, function ($mock) {
             $mock->shouldReceive('find')
                 ->andReturn(
-                    User::initModelWithData([
+                    new User([
+                        'id' => 1300005,
+                        'name_first' => 'Callum',
+                    ])
+                );
+            $mock->shouldReceive('initModelWithData')
```
Comment (AxonC) [inline tests/Unit/Endorsement/SpecialEndorsementRequestServiceTest.php:95]: Is this still current? I see you've removed the majority of the references to this method elsewhere.

**Review by AxonC — CHANGES_REQUESTED** [review]
@Keanu73 as a general note of advice, I'd try and keep the changes set as small as possible. There are a lot of changes here such as the L7 upgrade, changing the folder structures etc. which were not needed as part of this PR. It makes it much harder to review with such a large changeset & scope creep will almost always take over. 

Thank you for your work on this :)

**Comment by keanucz** [comment]
> @Keanu73 as a general note of advice, I'd try and keep the changes set as small as possible. There are a lot of changes here such as the L7 upgrade, changing the folder structures etc. which were not needed as part of this PR. It makes it much harder to review with such a large changeset & scope creep will almost always take over. 
> 
> Thank you for your work on this :)

I changed the folder structures just so everything would match with everything else. I will rollback the L7 upgrade, that is probably a tad too much.

**Comment by AxonC** [comment]
> 
> 
> > Also, any idea why codecov isn't working?
> 
> Doesn't seem to work outside of the main vatsim-uk repo for some reason. I think @AxonC was going to see if we could fix this?

The tokens are private to the parent repo. I'm not sure if there is an easy fix for this.

File: tests/Unit/Endorsement/SpecialEndorsementRequestServiceTest.php (hunk 12 above)
Comment (keanucz) [reply tests/Unit/Endorsement/SpecialEndorsementRequestServiceTest.php:95]: I was a bit inexperienced with Laravel a few days ago and one of the tests was failing with the initModelWithData mock, so I thought "ah let's just get rid of it, it'll work fine". I'll probably add back the references

File: app/Modules/Endorsement/Special/Services/SpecialEndorsementAssignment.php (hunk 10 above)
Comment (atoff) [reply app/Modules/Endorsement/Special/Services/SpecialEndorsementAssignment.php:25]: This looks like it was a change due to change in the Auth package. It implemented a macro on the BelongsTo relationship to add a contains method, so that we don't have to retrieve all the user's from the API. Ref: https://github.com/VATSIM-UK/auth-package/blob/master/src/UKAuthServiceProvider.php#L62

Comment (Keanu73) [reply app/Modules/Endorsement/Special/Services/SpecialEndorsementAssignment.php:25]: Can't really remember - I'll roll it back anyway

**Comment by AxonC** [comment]
> > @Keanu73 as a general note of advice, I'd try and keep the changes set as small as possible. There are a lot of changes here such as the L7 upgrade, changing the folder structures etc. which were not needed as part of this PR. It makes it much harder to review with such a large changeset & scope creep will almost always take over.
> > Thank you for your work on this :)
> 
> I changed the folder structures just so everything would match with everything else. I will rollback the L7 upgrade, that is probably a tad too much.

I'd like to get this merged - could you make these changes, please?

**Comment by keanucz** [comment]
> > > @Keanu73 as a general note of advice, I'd try and keep the changes set as small as possible. There are a lot of changes here such as the L7 upgrade, changing the folder structures etc. which were not needed as part of this PR. It makes it much harder to review with such a large changeset & scope creep will almost always take over.
> > > Thank you for your work on this :)
> > 
> > I changed the folder structures just so everything would match with everything else. I will rollback the L7 upgrade, that is probably a tad too much.
> 
> I'd like to get this merged - could you make these changes, please?

Yeah, I started working on them but my SSD busted itself, though I still have the changes on another  HDD so I'll have to copy them over and setup my dev environment again and whatnot

**Comment by keanucz** [comment]
Hrrrm.. for some reason the contains macro that you added isn't registering, any ideas @atoff?

**Comment by keanucz** [comment]
All good now - you now may review once more

**Comment by keanucz** [comment]
I also found a fix for the codecov issue, see https://github.com/VATSIM-UK/ukts-core-api/pull/57

**Review by AxonC — DISMISSED** [review]

**Comment by AxonC** [comment]
@Keanu73 there appears to be a test failure. Can you have a look?

**Comment by keanucz** [comment]
Looks like we have some race conditions on our hands..

**Comment by keanucz** [comment]
@AxonC Fixed the race conditions

**Review by AxonC — APPROVED** [review]

**Review by atoff — APPROVED** [review]
